### ISSUE TITLE
[APNG] UIImage RenderingMode is template only once.

### ISSUE THREAD

**Issue by 0k1019** [SDWebImage/SDWebImage#3665]
### New Issue Checklist

* [x ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 
 SDWebImage Version      | 5.18.7
 Integration Method      | SPM
 Xcode Version           | Xcode 15.1
 Repro rate              | only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

I have issue with first load APNG file load has `<SDAnimatedImage:0x28232ce70 anonymous {480, 400} renderingMode=automatic(original)>` So it render with [Template](https://developer.apple.com/documentation/uikit/uiimage/renderingmode/alwaystemplate) with tint color.
Can you share any options for Rendering Mode [original](https://developer.apple.com/documentation/uikit/uiimage/renderingmode/alwaysoriginal) always?

image path: https://apng.onevcat.com/assets/elephant.png
![image](https://github.com/SDWebImage/SDWebImage/assets/11485593/e123e397-a029-4048-8bcc-83e25a039379)

**Comment by dreampiggy** [comment]
We have no any design to support `tintColor` on Animated Image

**Comment by dreampiggy** [comment]
I don't think it's a common usage case.

Maybe this need much more work than expected. Because for `SDAnimatedImageView`, the render is hanlded by ourself, not UIImageView UIKit implementation.

We use layer to draw frame

**Comment by 0k1019** [comment]
Yes I know the layer architecture using SDAnimatedImagePlayer. 
My question is how can I render with `renderingMode alwaysOriginal`. 
It is not feature request i think...?
You mean It is now intended working in sdwebimage?

i just using options below,

```
SDImageCache.shared.clearMemory()
SDImageCache.shared.clearDisk()
SDWebImageContextOption.imageCoder
SDWebImageManager.shared.optionsProcessor = SDWebImageOptionsProcessor() { url, options, context in
      var mutableOptions = options
      if let context = context, let _ = context[.animatedImageClass] {
          mutableOptions.insert(.refreshCached)
      } else {
          mutableOptions.insert(.progressiveLoad)
          mutableOptions.insert(.refreshCached)
          
      }
      return SDWebImageOptionsResult(options: mutableOptions, context: context)
  }
```
and use this method `sd_setImage(with: url, placeholderImage: placeholderImage, options: [.retryFailed, .preloadAllFrames])`


I can load well in second time. as below image.
![IMG_60C531279AC2-1](https://github.com/SDWebImage/SDWebImage/assets/11485593/e9e02689-1b26-4f94-a8f4-b169ef75ace3)

**Comment by dreampiggy** [comment]
It should correctly render `alwaysOriginal` ?

Are you change the `UIImageView` (the super class)'s `tintColor` to non-nil value? Or using some traitCollection to override this

Check https://developer.apple.com/documentation/uikit/uiimageview/1621059-tintcolor?language=objc

**Comment by dreampiggy** [comment]
The decoded `UIImage` (or `SDAnimatedImage`, the same) actually use the `+[UIImage imageFromCGImage:]` to create.

Which inherit the UIKit's default value for `renderingMode`, which is `automatic`, by design.

We don't override the `renderingMode` of the SDWebImage decoded UIImage

**Comment by 0k1019** [comment]
Yes it should correctly render alwaysOriginal. I want to know the solution of it.
[this video](https://github.com/SDWebImage/SDWebImage/assets/11485593/b222f97e-d6a2-4bca-834a-6ecd946060b8) is my situation. it is abnormal. 
I set UIImageView `tintColor` to red for visualize issue. 
Above video, I didn't set `tintColor`
I know this library doesn't override renderingMode. But it happens to me...

**Comment by dreampiggy** [comment]
I guess your first appear rendering image is from 

[super setImage:image]

Since SDAnimatedImage is subclass, UIKit will try to something render.

And then, our player called back with the actual first frame (from APNG) and display directly on CALayer, then render as normal

**Comment by 0k1019** [comment]
It is so wired. I is not reproduce in simulator.
I check this library line by line. there is no rendering mode change code.
It seems to call super class funcions or initialize.
Can you check super class call code for resolve this issue?
Or any suggestions for reduce this problem reproduce?

**Comment by dreampiggy** [comment]
> It is so wired. It is not reproduce in simulator.

Sounds UIKit bug. Provide details about your test environment

OS version
iPhone model
SDWebImage version

**Comment by dreampiggy** [comment]
Actually I didn't find this issue on our Demo running simulator

---

A stupid quick hack I guess it to revert the  https://github.com/SDWebImage/SDWebImage/pull/3459

This will trigger a sync callback after first "setImage:" (after that PR we use async)

Since UIKit only render the last status in the same runloop, the sync callback will override the CALayer contents and should not show that temp tinted image poster.

But this will break SwiftUI integration. And I don't think it's a correct fix, but hide the issue
(there are still two image render result in the single "setImage:" call, wired)

**Comment by 0k1019** [comment]
My environment:
> OS version : iOS 16.0.2
> iPhone model: iPhone 14
> SDWebImage version: 5.18.7, 5.18.10

Thank you for sharing #3459 . 
After I change https://github.com/SDWebImage/SDWebImage/blob/59730af512c06fb569c119d737df4c1c499e001d/SDWebImage/Core/SDAnimatedImagePlayer.m#L121
to comment.
it is not reproduce my issue.

Can you check this line `self.needsDisplayWhenImageBecomesAvailable = YES;` is mandatory for `setupCurrentFrame`?
And If it is not mandatory, Can you remove this line in next release?

**Comment by dreampiggy** [comment]
No.

It hide the issue, but not fix the issue

**Comment by 0k1019** [comment]
Okay thank you for your consider about this issue.
Can you regenerate this issue?

**Comment by dreampiggy** [comment]
@0k1019 Provide a full demo (archive to zip and upload here)

I'll try to debug with.

Also, this only reproducable in Real Device ? Not simulator, correct ?
